#VRML_SIM R2021b utf8
WorldInfo {
  coordinateSystem "NUE"
}
Viewpoint {
  orientation -0.7605279026062298 0.6256701461869393 0.1735919857821294 0.6626546563521117
  position 1.0654767499197344 1.371225926760548 1.8000274420897922
}
TexturedBackground {
}
TexturedBackgroundLight {
}
RectangleArena {
}
DEF ACKERMAN Robot {
  translation 0 0.1 0.01
  children [
    DEF BODY_SHAPE Shape {
      appearance PBRAppearance {
        baseColor 0.8 0.8 0.8
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.4 0.1 0.2
      }
    }
    DEF REAR_RIGHT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor 0.15 0 -0.13
      }
      device [
        RotationalMotor {
          name "rear_right_wheel"
        }
      ]
      endPoint Solid {
        translation 0.15 0 -0.13
        rotation 0 0 1 0
        children [
          DEF WHEEL_ROT Transform {
            rotation 1 0 0 1.5708
            children [
              Shape {
                appearance PBRAppearance {
                  baseColor 0.8 0.8 0.8
                  roughness 1
                  metalness 0
                }
                geometry DEF WHEEL_CYL Cylinder {
                  height 0.04
                  radius 0.08
                }
              }
            ]
          }
        ]
        name "rear right wheel"
        boundingObject DEF WHEEL_ROT Transform {
          rotation 1 0 0 1.5708
          children [
            Shape {
              appearance PBRAppearance {
                baseColor 0.8 0.8 0.8
                roughness 1
                metalness 0
              }
              geometry DEF WHEEL_CYL Cylinder {
                height 0.04
                radius 0.08
              }
            }
          ]
        }
        physics Physics {
        }
      }
    }
    DEF REAR_LEFT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor 0.15 0 0.13
      }
      device [
        RotationalMotor {
          name "rear_left_wheel"
        }
      ]
      endPoint Solid {
        translation 0.15 0 0.13
        rotation 0 0 1 0
        children [
          USE WHEEL_ROT
        ]
        name "rear left wheel"
        boundingObject USE WHEEL_ROT
        physics Physics {
        }
      }
    }
    DEF RIGHT_STEER HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.15 0 -0.13
      }
      device [
        RotationalMotor {
          name "right_steer"
          minPosition -0.7
          maxPosition 0.7
        }
      ]
      endPoint Solid {
        translation -0.15 0 -0.13
        children [
          DEF FRONT_RIGHT_WHEEL HingeJoint {
            jointParameters HingeJointParameters {
              axis 0 0 1
            }
            device [
              RotationalMotor {
                name "front_right_wheel"
              }
            ]
            endPoint Solid {
              rotation 0 0 1 0
              children [
                USE WHEEL_ROT
              ]
              boundingObject USE WHEEL_ROT
              physics Physics {
              }
            }
          }
        ]
        name "right steer"
        boundingObject DEF DUMMY Sphere {
          radius 0.01
        }
        physics Physics {
        }
      }
    }
    DEF LEFT_STEER HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.15 0 0.13
      }
      device [
        RotationalMotor {
          name "left_steer"
          minPosition -0.7
          maxPosition 0.7
        }
      ]
      endPoint Solid {
        translation -0.15 0 0.13
        children [
          DEF FRONT_LEFT_WHEEL HingeJoint {
            jointParameters HingeJointParameters {
              axis 0 0 1
            }
            device [
              RotationalMotor {
                name "front_left_wheel"
              }
            ]
            endPoint Solid {
              rotation 0 0 1 0
              children [
                USE WHEEL_ROT
              ]
              boundingObject USE WHEEL_ROT
              physics Physics {
              }
            }
          }
        ]
        name "left steer"
        boundingObject DEF DUMMY Sphere {
          radius 0.01
        }
        physics Physics {
        }
      }
    }
  ]
  name "ackerman"
  boundingObject DEF BODY_SHAPE Shape {
    appearance PBRAppearance {
      baseColor 0.8 0.8 0.8
      roughness 1
      metalness 0
    }
    geometry Box {
      size 0.4 0.1 0.2
    }
  }
  physics Physics {
  }
  controller "robot_controller"
}
